openapi/openapi_v2_spec_analyser.go

```diff
@@ -282,63 +287,65 @@ func (specAnalyser *specV2Analyser) postDefined(resourceRootPath string) bool {
-	}
-	payloadDefinitionSchemaRef := bodyParameter.Schema
-	if payloadDefinitionSchemaRef == nil {
-		return "", fmt.Errorf("operation is missing the ref to the schema definition")
-	}
-	if payloadDefinitionSchemaRef.Ref.String() == "" {
-		return "", fmt.Errorf("operation has an invalid schema definition ref empty")
-	}
-	return payloadDefinitionSchemaRef.Ref.String(), nil
-}
 
-// getPayloadDefName only supports references to the same document. External references like URLs is not supported at the moment
-func (specAnalyser *specV2Analyser) getPayloadDefName(ref string) (string, error) {
-	reg, err := regexp.Compile(swaggerResourcePayloadDefinitionRegex)
-	if err != nil {
-		return "", fmt.Errorf("an error occurred while compiling the swaggerResourcePayloadDefinitionRegex regex '%s': %s", swaggerResourcePayloadDefinitionRegex, err)
-	}
-	payloadDefName := reg.FindStringSubmatch(ref)[0]
-	if payloadDefName == "" {
-		return "", fmt.Errorf("could not find a valid definition name for '%s'", ref)
-	}
-	return payloadDefName, nil
+	if bodyCounter <= 0 {
+		return nil, fmt.Errorf("resource root operation missing the body parameter")
+	}
+
+	if bodyCounter > 1 {
+		return nil, fmt.Errorf("resource root operation contains multiple 'body' parameters")
+	}
+
+	if bodyParameter.Schema == nil {
+		return nil, fmt.Errorf("resource root operation missing the schema for the POST operation body parameter")
+	}
+
+	if bodyParameter.Schema.Ref.String() != "" {
+		return nil, fmt.Errorf("the operation ref was not expanded properly, check that the ref is valid (no cycles, bogus, etc)")
+	}
+
+	// This means that either:
+	// - the schema has been expanded OR
+	// - the parameter has the schema embedded (this is not recommended since that might lead to users defining the
+	// embedded schema only considering the parameters for the POST operation and not the computed values that may be
+	// returned by the GET operation resulting into a miss-configured terraform resource. Example:
+	// paths:
+	//  /v1/cdns:
+	//     post:
+	//      summary: Create cdn
+	//      parameters:
+	//        - in: body
+	//          name: body
+	//          description: Created CDN
+	//          schema:
+	//            type: object
+	//            properties:
+	//              name:
+	//                type: string
+	//              last_name:
+	//                type: string
+	//
+	// if the GET operation for the /v1/cdns/{id} path includes readOnly properties terraform will find diffs and fail at runtime
+	// Hence, the previous enforcement on using $ref pointing at a common model definition used across the different resource operations (POST, GET, PUT).
+	// Considering the expansion now happens before this point, there is no way to identify whether the the embedded schema
+	// is result of the expansion (in which case this should work fine, since the expansion was created from a ref link) or the user actually
+	// defining the model embedded.
```

Comment:
I think we should move this to the documentation instead of leaving it here.

Reply:
+1